src/runtime-rs/crates/hypervisor/src/qemu/inner.rs

```diff
@@ -86,26 +159,52 @@ impl QemuInner {
 
     pub(crate) async fn get_thread_ids(&self) -> Result<VcpuThreadIds> {
         info!(sl!(), "QemuInner::get_thread_ids()");
-        todo!()
+        //todo!()
+        let vcpu_thread_ids: VcpuThreadIds = VcpuThreadIds {
+            vcpus: HashMap::new(),
+        };
+        Ok(vcpu_thread_ids)
     }
 
     pub(crate) async fn get_vmm_master_tid(&self) -> Result<u32> {
         info!(sl!(), "QemuInner::get_vmm_master_tid()");
-        todo!()
+        if let Some(qemu_process) = &self.qemu_process {
+            info!(
+                sl!(),
+                "QemuInner::get_vmm_master_tid(): returning {}",
+                qemu_process.id()
+            );
+            Ok(qemu_process.id())
+        } else {
+            Err(anyhow!("qemu process not running"))
+        }
     }
 
     pub(crate) async fn get_ns_path(&self) -> Result<String> {
         info!(sl!(), "QemuInner::get_ns_path()");
-        todo!()
+        //todo!()
+        Ok(format!(
+            "/proc/{}/task/{}/ns",
+            std::process::id(),
+            std::process::id()
+        ))
     }
 
     pub(crate) async fn cleanup(&self) -> Result<()> {
         info!(sl!(), "QemuInner::cleanup()");
-        todo!()
+        let vm_path = [KATA_PATH, self.id.as_str()].join("/");
+        std::fs::remove_dir_all(vm_path)?;
+        Ok(())
     }
 
-    pub(crate) async fn resize_vcpu(&self, _old_vcpus: u32, _new_vcpus: u32) -> Result<(u32, u32)> {
-        info!(sl!(), "QemuInner::resize_vcpu()");
+    pub(crate) async fn resize_vcpu(&self, old_vcpus: u32, new_vcpus: u32) -> Result<(u32, u32)> {
+        info!(
+            sl!(),
+            "QemuInner::resize_vcpu(): {} -> {}", old_vcpus, new_vcpus
+        );
+        if new_vcpus == old_vcpus {
+            return Ok((old_vcpus, new_vcpus));
+        }
         todo!()
```

Comment:
My rationale for this was that apart from the trivial case, the actual implementation of the function is literally a TODO.  As a developer I'd rather be reminded of this by a hard panic than having to comb through logs. Also I suspect that an Err return value from this function is meant to convey different meanings so I thought returning an Err could be misleading.

I do agree though that it looks odd and if there's a strong opinion in favour of returning Err I can do that.

Reply:
Sure that macro is useful for development, but does that mean this code should not be used generally yet? Do you plan to remove these panic calls on a later PR to avoid runtime panics using runtime-rs and QEMU?

Reply:
Yes, exactly, this PR just lays down a framework and adds some very basic functionality.  The QEMU driver is still very much a WIP at this stage and should not be used generally yet.  A proper implementation of this function will definitely be added later.

The purpose of this PR is 1) to present the driver in a reasonably-sized chunks that are arguably easier to digest and review than a huge PR with a complete driver, 2) to let the community see which direction the driver implementation is moving in and give feedback early if necessary, and 3) for me to avoid repeated conflicts on pulling from upstream which are unavoidable when a big chunk of code is developed in isolation for a long period of time while the rest of runtime-rs keeps changing fast.